src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    /// alignment of `layout`, or if reallocation otherwise fails. (Note
+    /// that did not say "if and only if" -- in particular, an
+    /// implementation of this method *can* return `Ok` if
+    /// `new_layout.align() != old_layout.align()`; or it can return `Err`
+    /// in that scenario, depending on whether this allocator
+    /// can dynamically adjust the alignment constraint for the block.)
+    ///
+    /// If this method returns `Err`, then ownership of the memory
+    /// block has not been transferred to this allocator, and the
+    /// contents of the memory block are unaltered.
+    unsafe fn realloc(&mut self,
+                      ptr: Address,
+                      layout: Layout,
+                      new_layout: Layout) -> Result<Address, AllocErr> {
+        let (min, max) = self.usable_size(&layout);
+        let s = new_layout.size();
+        // All Layout alignments are powers of two, so a comparison
+        // suffices here (rather than resorting to a `%` operation).
+        if min <= s && s <= max && new_layout.align() <= layout.align() {
+            return Ok(ptr);
+        } else {
+            let new_size = new_layout.size();
+            let old_size = layout.size();
+            let result = self.alloc(new_layout);
+            if let Ok(new_ptr) = result {
+                ptr::copy(ptr as *const u8, new_ptr, cmp::min(old_size, new_size));
+                self.dealloc(ptr, layout);
+            }
+            result
+        }
+    }
+
+    /// Behaves like `fn alloc`, but also ensures that the contents
+    /// are set to zero before being returned.
+    unsafe fn alloc_zeroed(&mut self, layout: Layout) -> Result<Address, AllocErr> {
+        let size = layout.size();
+        if size > ::core::isize::MAX as usize {
+            return Err(AllocErr::Unsupported { details: "isize overflow" });
+        }
+        let size = size as isize;
+        let p = self.alloc(layout);
+        if let Ok(p) = p {
+            for i in 0..size {
+                *p.offset(i) = 0;
+            }
+        }
+        p
+    }
+
+    /// Behaves like `fn alloc`, but also returns the whole size of
+    /// the returned block. For some `layout` inputs, like arrays, this
+    /// may include extra storage usable for additional data.
+    unsafe fn alloc_excess(&mut self, layout: Layout) -> Result<Excess, AllocErr> {
+        let usable_size = self.usable_size(&layout);
+        self.alloc(layout).map(|p| Excess(p, usable_size.1))
+    }
+
+    /// Behaves like `fn realloc`, but also returns the whole size of
+    /// the returned block. For some `layout` inputs, like arrays, this
+    /// may include extra storage usable for additional data.
+    unsafe fn realloc_excess(&mut self,
+                             ptr: Address,
+                             layout: Layout,
+                             new_layout: Layout) -> Result<Excess, AllocErr> {
+        let usable_size = self.usable_size(&new_layout);
```

Comment:
yeah I think this would only win if you were hooking in an allocator that *both*
 1. overrides `fn alloc_excess`, and
 2. does *not* override `fn realloc_excess`.

Probably better to just advise allocator implementators to understand what the default implementations do when deciding which methods to override.